torch/distributed/_composable/fsdp/_fsdp_common.py

```diff
@@ -117,20 +117,29 @@ def _from_local_no_grad(
     global_stride: Tuple[int, ...],
 ) -> DTensor:
     """
-    This method is similar to ``DTensor.from_local()`` except it avoids some
-    CPU overhead by avoiding default args and not being differentiable.
+    This method is similar to ``DTensor.from_local()`` except that in eager mode
+    it avoids some CPU overhead by avoiding default args and not being differentiable.
     """
-    return DTensor(
-        # Use the local tensor directly instead of constructing a new tensor
-        # variable, e.g. with `view_as()`, since this is not differentiable
-        local_tensor,
-        device_mesh,
-        placements,
-        shape=global_size,
-        dtype=local_tensor.dtype,
-        requires_grad=local_tensor.requires_grad,
-        stride=global_stride,
-    )
+    if not torch._dynamo.compiled_autograd.compiled_autograd_enabled:
```

Comment:
Actually, I wonder if we can do `if not torch.is_grad_enabled():`?

I think `_from_local_no_grad` on the hot path is mainly for converting sharded gradient to `DTensor` in `foreach_reduce`, which is outside autograd (since `@torch.no_grad()` decorator).

Reply:
I found the limitation is that compile cannot support DTensor constructor either in grad mode or no-grad mode, so we really need a separate path for compile :(

Hopefully we fix these tensor subclass papercuts very soon and can remove this divergence.

Reply:
Sounds good!